### PR TITLE
Handeharputlu assignment3

### PR DESCRIPTION
**Pull request by handeharputlu**

### CODE REVIEW

File: handeharputlu/assignment-3/Lexicon.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+	// Lexicon constructs a lexicon whose wordList consists of elements of the
+	// argument,
+	// only transformed to lower cases to facilitate comparisons of strings and
+	// implementations other methods
+	public Lexicon(ArrayList<String> wordList) {
+
+		assert (wordList != null);
+		this.wordList = wordList;
+		for (int i = 0; i < wordList.size(); i++)
```
Comment (ltache) [inline handeharputlu/assignment-3/Lexicon.java]: I'd suggest using brackets even if there's one single instruction. Omitting braces harms readability and can easily cause issues when the code is modified. For example, let's say you are temporarily commenting out code to debug something:

`if(condition) 
//      statement;
otherStatement;`

File: handeharputlu/assignment-3/Lexicon.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+public class Lexicon {
+
+	private ArrayList<String> wordList;
+
+	public Lexicon() {
+		this.wordList = new ArrayList<String>();
+	}
+
+	// Lexicon constructs a lexicon whose wordList consists of elements of the
+	// argument,
```
Comment (ltache) [inline handeharputlu/assignment-3/Lexicon.java]: Nit: Here seems to be some extra spaces and a missing period at the end of the sentence.

File: handeharputlu/assignment-3/Lexicon.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+			throw new IllegalArgumentException("Empty word isn't acceptable");
+		}
+
+		return this.wordList.contains(word.toLowerCase());
+	}
+
+	// isPrefix searches if given word is a prefix of any word in the lexicon.To do
+	// this, it iterates
+	// over wordList of lexicon and check if it is prefix until a match is found or
+	// the word
```
Comment (ltache) [inline handeharputlu/assignment-3/Lexicon.java]: These 2 lines can be merged together (42-43). Seems there are some extra spaces in the comment.

File: handeharputlu/assignment-3/Lexicon.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	// isWord searches if given word exists in the lexicon. To do this, it iterates
+	// over wordList of lexicon and compare words until a match is found or the word
+	// doesn't exist in lexicon.
+
+	public boolean isWord(String word) {
+		assert (word != null);
+		if (word.isEmpty()) {
+			throw new IllegalArgumentException("Empty word isn't acceptable");
```
Comment (ltache) [inline handeharputlu/assignment-3/Lexicon.java]: Why do we have this exception here? It doesn't seem to be a corner-case for the next line (36) and this.wordList.contains will return false if the empty string is not in the lexicon. WDYT?

File: handeharputlu/assignment-3/Lexicon.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+	// isPrefix searches if given word is a prefix of any word in the lexicon.To do
+	// this, it iterates
+	// over wordList of lexicon and check if it is prefix until a match is found or
+	// the word
+	// isn't a prefix of another in lexicon.
+	public boolean isPrefix(String word) {
+		assert (word != null);
+		if (word.isEmpty()) {
+			throw new IllegalArgumentException("Empty word isn't acceptable");
```
Comment (ltache) [inline handeharputlu/assignment-3/Lexicon.java]: Same as the above.

File: handeharputlu/assignment-3/Lexicon.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		}
+
+		for (int i = 0; i < wordList.size(); i++) {
+			if (wordList.get(i).contains(word.toLowerCase())) {
```
Comment (ltache) [inline handeharputlu/assignment-3/Lexicon.java]: The contains() method is Java method to check if String contains another substring or not. For example for the word "cart" and the string "art" it will return true. However, we can use the method startsWith() which returns boolean value true or false based on whether the given string begins with the specified letter or word.

File: handeharputlu/assignment-3/Test_Lexicon.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+		assertEquals(temp.size(), temp2.size());
+		assertEquals("ann", temp2.get(0));
+		assertEquals("bad", temp2.get(1));
+		assertEquals("good", temp2.get(2));
+		assertEquals("twenty", temp2.get(3));
+
+	}
+
+	@Test
+	void IsWordDoesntWordWithNotInitializedWord() {
```
Comment (ltache) [inline handeharputlu/assignment-3/Test_Lexicon.java]: Nit: IsWordDoesntWorkWithNotInitializedWord

File: handeharputlu/assignment-3/Test_Lexicon.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+			lex.isWord(s);
+		} catch (AssertionError e) {
+			thrown = true;
+		}
+
+		assertTrue(thrown);
+	}
+
+	@Test
+	void IsWordDoesntWordWithEmptyWord() {
```
Comment (ltache) [inline handeharputlu/assignment-3/Test_Lexicon.java]: I'm not sure if this is actually a problem TBH.

File: handeharputlu/assignment-3/Test_Lexicon.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+			lex.isPrefix(s);
+		} catch (AssertionError e) {
+			thrown = true;
+		}
+
+		assertTrue(thrown);
+	}
+
+	@Test
+	void IsPrefixDoesntWorkWithEmptyWord() {
```
Comment (ltache) [inline handeharputlu/assignment-3/Test_Lexicon.java]: I'm not sure if this is actually a problem TBH. I think we should follow the pattern of the StartWith method I mentioned before. However, as the empty string is prefix for any other string I'd suggest we just return true.

File: handeharputlu/assignment-3/Test_Lexicon.java [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+			if (e.getMessage() == "Empty word isn't acceptable") {
+				thrown = true;
+			}
+		}
+
+		assertTrue(thrown);
+	}
+
+	@Test
+	void IsPrefixWorks() {
```
Comment (ltache) [inline handeharputlu/assignment-3/Test_Lexicon.java]: Taking into consideration that you converted all the string to lower case this test is a bit confusing. Maybe it's more intuitive to make all the words upper case (similar with what we have in the problem's statement) or lower case. 
 
Please add some tests for some more complicated cases like word = "TWENTY" and check isPrefix("NTY"). Adding a big variety of tests we'll help us to spot error like using the wrong method or hidden bugs in the code.

File: handeharputlu/assignment-3/WordSearch.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+	// to lexicon.To make this search, grid should be of dimension of at least 1x1.
+	// To find all the words in grid, we simply do a depth first search for each
+	// letter in grid and check neighbors recursively until all the possible words
+	// starting with this letter is found.The search for each letter is stopped
+	// until all the possible paths in grid are found and checked
+	// or developing a branch of the current word is stopped if isn't a prefix of
+	// any others.
+	public static ArrayList<String> findWords(char[][] grid, Lexicon lexicon) {
+
+		assert grid != null : "grid can't be null";
```
Comment (ltache) [inline handeharputlu/assignment-3/WordSearch.java]: I think it's depends on the design but usually we don't want to assert for these corner-cases. The best practice is quietly treat these corner-cases without breaking the program. Let's imagine we have a pipeline running multiple tasks in production. If by any reason we receive null as input we don't want to break everything but rather return null instead and move further. I'd suggest to check these corner-cases with ifs:

```java
if (grid == null) {
    return new ArrayList<>();
}
```

File: handeharputlu/assignment-3/WordSearch.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+		assert grid.length >= 1 : "grid has to have at least 1 row";
+		assert grid[0] != null : "grid's second dimension can't be null";
+		assert grid[0].length >= 1 : "grid has to have at least 1 column";
+		assert (lexicon != null) : "lexicon can't be null";
+
+		int widthGrid = grid.length;
+		int heightGrid = grid[0].length;
+
+		boolean[][] visited = new boolean[widthGrid][heightGrid];
+		ArrayList<String> wordsFound = new ArrayList<String>();
```
Comment (ltache) [inline handeharputlu/assignment-3/WordSearch.java]: Nit & nice to have: Since Java 8 we can specify the type only once:

```java
ArrayList<String> wordsFound = new ArrayList<>();
```

File: handeharputlu/assignment-3/WordSearch.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+		return wordsFound;
+	}
+
+	private static void findWordsLexicon(char grid[][], boolean visited[][], int i, int j, String s,
+			ArrayList<String> wordsFound, Lexicon lexicon) {
+
+		visited[i][j] = true;
+		s = s + grid[i][j];
+
+		if (lexicon.isWord(s) && !wordsFound.contains(s.toLowerCase())) {
```
Comment (ltache) [inline handeharputlu/assignment-3/WordSearch.java]: We can use a Set instead of an array. `wordsFound.contains(s)` will take O(n) times on an array because it will iterate through all the words in a container. On the other hand, checking if a word is in a set is done in constant time O(1).

File: handeharputlu/assignment-3/WordSearch.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+		for (int row = i - 1; row <= i + 1 && row < grid.length; row++) {
+			for (int col = j - 1; col <= j + 1 && col < grid[0].length; col++) {
+				if (row >= 0 && col >= 0 && !visited[row][col] && lexicon.isPrefix(s)) {
+					findWordsLexicon(grid, visited, row, col, s, wordsFound, lexicon);
+				}
+			}
+		}
+
+		s = "" + s.charAt(s.length() - 1);
```
Comment (ltache) [inline handeharputlu/assignment-3/WordSearch.java]: This line it's a bit unclear to me. Can you please clarify? Seems like only the second to last character from the string is assigned to s. Is this intended?

File: handeharputlu/assignment-3/WordSearch.java [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+		visited[i][j] = true;
+		s = s + grid[i][j];
+
+		if (lexicon.isWord(s) && !wordsFound.contains(s.toLowerCase())) {
+			wordsFound.add(s.toLowerCase());
+		}
+
+		for (int row = i - 1; row <= i + 1 && row < grid.length; row++) {
+			for (int col = j - 1; col <= j + 1 && col < grid[0].length; col++) {
+				if (row >= 0 && col >= 0 && !visited[row][col] && lexicon.isPrefix(s)) {
```
Comment (ltache) [inline handeharputlu/assignment-3/WordSearch.java]: Taking into consideration that s is not changes since line 43 I'd suggest we check if it's prefix immediately after that so that we don't do all these iterations if it's not prefix.

File: handeharputlu/assignment-3/Test_WordSearch.java [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+		}
+		assertTrue(thrown);
+	}
+
+	@Test
+	void wordSearchWorks() {
+		ArrayList<String> temp = new ArrayList<String>();
+		temp.add("CAR");
+		temp.add("CARD");
+		temp.add("cArT");
```
Comment (ltache) [inline handeharputlu/assignment-3/Test_WordSearch.java]: Same as the above. We can add tests only with lower/upper case since we convert the strings to lower case anyway.

File: handeharputlu/assignment-3/Lexicon.java [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+	// over wordList of lexicon and check if it is prefix until a match is found or
+	// the word
+	// isn't a prefix of another in lexicon.
+	public boolean isPrefix(String word) {
+		assert (word != null);
+		if (word.isEmpty()) {
+			throw new IllegalArgumentException("Empty word isn't acceptable");
+		}
+
+		for (int i = 0; i < wordList.size(); i++) {
```
Comment (ltache) [inline handeharputlu/assignment-3/Lexicon.java]: An optimisation which can be done here is to make the string lower case before the for loop. By doing that we'll do the operation only once compared with the current situation when we do it wordList.size() times.

File: handeharputlu/assignment-3/Lexicon.java [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+			if (wordList.get(i).contains(word.toLowerCase())) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	public ArrayList<String> getWordList() {
```
Comment (ltache) [inline handeharputlu/assignment-3/Lexicon.java]: I don't see where this function is used (besides the tests). In general it's a good idea to test that the Lexicon contains all the words you've added but if the method returns a list of words you cannot assume in the test that the words are sorted in some specific order, so you should maybe adapt the check in the test.

File: handeharputlu/assignment-3/Lexicon.java (hunk 4 above)
Comment (handeharputlu) [reply handeharputlu/assignment-3/Lexicon.java]: I agree with you. I added them to have an idea about the exception in my tests . Because in that method, there are many corner cases to test. To understand which one is problematic, I needed an IllegalArgumentException with a message and thanks to the message, I could track the problems about the code.

File: handeharputlu/assignment-3/Lexicon.java (hunk 5 above)
Comment (handeharputlu) [reply handeharputlu/assignment-3/Lexicon.java]: Same as the above too

File: handeharputlu/assignment-3/Lexicon.java (hunk 6 above)
Comment (handeharputlu) [reply handeharputlu/assignment-3/Lexicon.java]: That is a much better way to do it, thank you!

File: handeharputlu/assignment-3/Test_Lexicon.java (hunk 8 above)
Comment (handeharputlu) [reply handeharputlu/assignment-3/Test_Lexicon.java]: Can you clarify it a bit more? For me an empty string is a substring of any others, so this may be a problem.

File: handeharputlu/assignment-3/Lexicon.java (hunk 18 above)
Comment (handeharputlu) [reply handeharputlu/assignment-3/Lexicon.java]: Yes, I only included it for the tests.So, you mean that I should sort the list of words before compare them?

File: handeharputlu/assignment-3/Test_Lexicon.java (hunk 10 above)
Comment (handeharputlu) [reply handeharputlu/assignment-3/Test_Lexicon.java]: I am not sure if I understand your first paragraph. Can you clarify it a little bit more?

Comment (ltache) [reply handeharputlu/assignment-3/Test_Lexicon.java]: Clarified offline. I was saying that here is not context about the fact that in the method everything is converted to lower case and it might be confusing.

File: handeharputlu/assignment-3/Lexicon.java (hunk 18 above)
Comment (ltache) [reply handeharputlu/assignment-3/Lexicon.java]: Clarified offline. I would say there are 2 options: sort the strings in the test method or check with contains() rather than having the element at a particular position.
